File: caesar/src/Strategies.hs
module Strategies where

import qualified Data.Map as Map
import Data.List (sortOn)
import Data.Ord (Down(Down))
import Metadata
import DataRetriever

data PlacementStrategy = HOT_TABLE | HOT_COLUMN | LOPT deriving (Show, Eq)

data Placement = Placement
  { placementColumnId :: Int
  , placementDeviceId :: Int
  } deriving (Show, Eq)

hotTableStrategy :: Metadata -> [Placement]
hotTableStrategy meta =
  let tableAccessCounts = foldr (\entry m -> Map.insertWith (+) (traceTableId entry) 1 m) Map.empty (metaTrace meta)
      sortedTables = sortOn (Down . snd) $ Map.toList tableAccessCounts
      sortedDevices = sortOn (Down . deviceCostPerGB) $ Map.elems (metaDevices meta)
  in concat
     [ [ Placement (columnId col) (deviceId dev)
       | col <- tableColumns tbl
       ]
     | ((tblId, _), dev) <- zip sortedTables sortedDevices
     , Just tbl <- [getTable tblId meta]
     ]

hotColumnStrategy :: Metadata -> [Placement]
hotColumnStrategy meta =
  let columnAccessCounts = foldr (\entry m -> foldr (\colId n -> Map.insertWith (+) colId 1 n) m (traceColumnIds entry)) Map.empty (metaTrace meta)
      sortedColumns = sortOn (Down . snd) $ Map.toList columnAccessCounts
      sortedDevices = sortOn (Down . deviceCostPerGB) $ Map.elems (metaDevices meta)
  in [ Placement colId (deviceId dev)
     | ((colId, _), dev) <- zip sortedColumns sortedDevices
     ]

loptStrategy :: Maybe Float -> Metadata -> [Placement]
loptStrategy budget meta =
  let columnSizes = Map.map columnSize (metaColumns meta)
      sortedColumns = sortOn (Down . snd) $ Map.toList columnSizes
      sortedDevices = sortOn deviceCostPerGB $ Map.elems (metaDevices meta)
      
      place _ _ [] _ = []
      place [] _ _ _ = []
      place ((colId, size):restCols) (dev:restDevs) acc totalCost
        | Just b <- budget, newCost > b = place ((colId, size):restCols) restDevs acc totalCost
        | otherwise = Placement colId (deviceId dev) : place restCols (dev:restDevs) (newCost:acc) newCost
        where newCost = totalCost + (fromIntegral size / (1024 * 1024 * 1024)) * deviceCostPerGB dev
      
  in place sortedColumns sortedDevices [] 0

placeData :: PlacementStrategy -> Maybe Float -> Metadata -> [Placement]
placeData HOT_TABLE _ meta = hotTableStrategy meta
placeData HOT_COLUMN _ meta = hotColumnStrategy meta
placeData LOPT budget meta = loptStrategy budget meta

predictScanTime :: [Placement] -> TraceEntry -> Metadata -> Float
predictScanTime placements tableScan meta =
  let deviceTimes = foldr (\colId acc ->
        let placement = find (\p -> placementColumnId p == colId) placements
            device = placement >>= \p -> getDevice (placementDeviceId p) meta
            devModel = placement >>= \p -> getDeviceModel (placementDeviceId p) meta
            column = getColumn colId meta
        in case (device, devModel, column) of
             (Just dev, Just model, Just col) ->
               let compressionThroughput = case deviceCompression dev of
                     None -> modelNoneThroughput model
                     LZ4 -> modelLZ4Throughput model
                     ZSTD -> modelZSTDThroughput model
                   columnTime = modelSeekTime model + (fromIntegral (columnSize col) / compressionThroughput)
               in Map.insertWith (+) (deviceId dev) columnTime acc
             _ -> acc
      ) Map.empty (traceColumnIds tableScan)
  in maximum (Map.elems deviceTimes)

predictTotalTime :: [Placement] -> Metadata -> Float
predictTotalTime placements meta =
  sum [predictScanTime placements trace meta | trace <- metaTrace meta]

calculateTotalCost :: [Placement] -> Metadata -> Float
calculateTotalCost placements meta =
  sum [deviceCostPerGB dev * (fromIntegral (columnSize col) / (1024 * 1024 * 1024))
      | Placement colId devId <- placements
      , Just dev <- [getDevice devId meta]
      , Just col <- [getColumn colId meta]
      ]

optimizePlacement :: Float -> Metadata -> IO [Placement]
optimizePlacement budget meta = do
  let placements = placeData LOPT (Just budget) meta
      totalTime = predictTotalTime placements meta
      totalCost = calculateTotalCost placements meta
  putStrLn $ "Predicted total time: " ++ show totalTime ++ " s"
  putStrLn $ "Total cost: " ++ show totalCost
  return placements

applyPlacements :: [Placement] -> Metadata -> KVStore -> DataRetriever -> IO (KVStore, DataRetriever)
applyPlacements placements meta kvstore retriever =
  foldM (\(kv, ret) placement -> do
    let colId = placementColumnId placement
        devId = placementDeviceId placement
        chunks = Map.findWithDefault [] colId (retrieverColumnChunks ret)
    (newKV, newRet) <- foldM (\(kv', ret') chunk ->
      if chunkDeviceId chunk /= devId
      then do
        let (updatedKV, updatedRet) = moveColumnChunk chunk devId meta kv' ret'
        putStrLn $ "Moved column " ++ show colId ++ " to device " ++ show devId
        return (updatedKV, updatedRet)
      else return (kv', ret')
      ) (kv, ret) chunks
    return (newKV, newRet)
  ) (kvstore, retriever) placements

File: anna/src/strategies.ml
open Metadata
open Prediction_model
open Lopt_solver

type placement_strategy = HOT_TABLE | HOT_COLUMN | LOPT

let hot_table_strategy () : placement list =
  let table_access_counts = Hashtbl.create 10 in
  List.iter (fun entry ->
    let count = Hashtbl.find_opt table_access_counts entry.table_id
                |> Option.value ~default:0 in
    Hashtbl.replace table_access_counts entry.table_id (count + 1)
  ) !trace;

  let sorted_tables = Hashtbl.to_seq table_access_counts
                      |> List.of_seq
                      |> List.sort (fun (_, c1) (_, c2) -> compare c2 c1) in

  let sorted_devices = Hashtbl.to_seq devices
                       |> List.of_seq
                       |> List.sort (fun (_, d1) (_, d2) -> compare d2.cost_per_gb d1.cost_per_gb) in

  let rec assign_tables tables devices acc =
    match tables, devices with
    | [], _ | _, [] -> acc
    | (table_id, _) :: rest_tables, (device_id, _) :: rest_devices ->
      let table = Hashtbl.find tables table_id in
      let new_placements = List.map (fun col -> 
        {column_id = col.id; device_id = device_id}
      ) table.columns in
      assign_tables rest_tables rest_devices (acc @ new_placements)
  in

  assign_tables sorted_tables sorted_devices []

let hot_column_strategy () : placement list =
  let column_access_counts = Hashtbl.create 1000 in
  List.iter (fun entry ->
    List.iter (fun col_id ->
      let count = Hashtbl.find_opt column_access_counts col_id
                  |> Option.value ~default:0 in
      Hashtbl.replace column_access_counts col_id (count + 1)
    ) entry.column_ids
  ) !trace;

  let sorted_columns = Hashtbl.to_seq column_access_counts
                       |> List.of_seq
                       |> List.sort (fun (_, c1) (_, c2) -> compare c2 c1) in

  let sorted_devices = Hashtbl.to_seq devices
                       |> List.of_seq
                       |> List.sort (fun (_, d1) (_, d2) -> compare d2.cost_per_gb d1.cost_per_gb) in

  let rec assign_columns columns devices acc =
    match columns, devices with
    | [], _ | _, [] -> acc
    | (column_id, _) :: rest_columns, (device_id, _) :: rest_devices ->
      let new_placement = {column_id = column_id; device_id = device_id} in
      assign_columns rest_columns rest_devices (new_placement :: acc)
  in

  assign_columns sorted_columns sorted_devices []

let lopt_strategy (budget: float option) : placement list =
  solve_lopt budget

let place_data (strategy: placement_strategy) (budget: float option) : placement list =
  match strategy with
  | HOT_TABLE -> hot_table_strategy ()
  | HOT_COLUMN -> hot_column_strategy ()
  | LOPT -> lopt_strategy budget

let optimize_placement (budget: float) : placement list =
  let placements = place_data LOPT (Some budget) in
  let total_time = predict_total_time placements !trace in
  let total_cost = calculate_total_cost placements in
  Printf.printf "Predicted total time: %.2f s\n" total_time;
  Printf.printf "Total cost: %.2f\n" total_cost;
  placements

let apply_placements (placements: placement list) : unit =
  List.iter (fun placement ->
    let column = Hashtbl.find columns placement.column_id in
    let target_device = Hashtbl.find devices placement.device_id in
    let chunks = Hashtbl.find Data_retriever.column_chunks placement.column_id in
    
    List.iter (fun chunk ->
      if chunk.device_id <> placement.device_id then
        Data_retriever.move_column_chunk chunk placement.device_id
    ) chunks;
    
    Printf.printf "Moved column %s to device %s\n" column.name target_device.name;
  ) placements